tools/approval-service/internal/approvalservice/eventprocessor/githubprocessor/githubhandler.go

```diff
@@ -0,0 +1,249 @@
+		return nil, fmt.Errorf("getting workflow run info: %w", err)
+	}
+
+	reason, err := genReason(e, runInfo)
+	if err != nil {
+		return nil, fmt.Errorf("generating reason template: %w", err)
+	}
+	req.SetRequestReason(reason)
+	err = p.store.StoreWorkflowInfo(ctx, req, store.GitHubWorkflowInfo{
+		Org:           e.Organization,
+		Repo:          e.Repository,
+		Env:           e.Environment,
+		WorkflowRunID: e.WorkflowID,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("storing data: %w", err)
+	}
+
+	newReq, err := p.teleClient.CreateAccessRequestV2(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("creating access request: %w", err)
+	}
+
+	_, denyErr := p.automaticallyDenied(e)
+	if denyErr != nil {
+		// TODO: Consider auto denying the request here.
+		return nil, fmt.Errorf("automatically denied: %w", denyErr)
+	}
+
+	return newReq, nil
+}
+
+// Performs approval checks that are GH-specific. This should only be used to deny requests,
+// never approve them.
+func (p *Processor) automaticallyDenied(e githubevents.DeploymentReviewEvent) (deny bool, err error) {
+	if e.Organization != p.org {
+		return true, fmt.Errorf("organization %q does not match expected organization %q", e.Organization, p.org)
+	}
+
+	if e.Repository != p.repo {
+		return true, fmt.Errorf("repository %q does not match expected repository %q", e.Repository, p.repo)
+	}
+
+	if !slices.Contains(p.envs, e.Environment) {
+		return true, fmt.Errorf("environment %q does not match expected environments %q", e.Environment, p.envs)
+	}
+
+	// TODO: check user is part of one of valid orgs
+
+	return false, nil
+}
+
+// HandleReview handles updates to the state of a Teleport Access Request and updates the state of the GitHub deployment accordingly.
+// This implements the [accessrequest.ReviewHandler] interface.
+func (p *Processor) HandleReview(ctx context.Context, req types.AccessRequest) error {
+	p.log.Info("Handling review", "access_request_name", req.GetName())
+
+	data, err := p.store.GetWorkflowInfo(ctx, req)
+	if err != nil {
+		return fmt.Errorf("getting data: %w", err)
+	}
+
+	var state github.PendingDeploymentApprovalState
+	switch req.GetState() {
+	case types.RequestState_APPROVED:
+		state = github.PendingDeploymentApprovalStateApproved
+	case types.RequestState_DENIED:
+		state = github.PendingDeploymentApprovalStateRejected
+	default:
+		return fmt.Errorf("invalid state %q for access request %q", req.GetState(), req.GetName())
+	}
+
+	p.log.Info("Updating deployment", "access_request_name", req.GetName(), "environment", data.Env, "state", state)
+	if err := p.client.ReviewDeploymentProtectionRule(ctx, data.Org, data.Repo, data.WorkflowRunID, state, data.Env,
+		fmt.Sprintf("Access by pipeline approval service. Access Request ID: %s", req.GetName())); err != nil {
+		return fmt.Errorf("reviewing deployment protection rule: %w", err)
+	}
+
+	return nil
+}
+
+var reasonTmpl = template.Must(template.New("reason").Parse(`GitHub Deployment Review for:
```

Comment:
These singleton templates do not need a name to `New()` - I always just pass an empty string as that name has no meaning.